volatility3/framework/plugins/windows/direct_system_calls.py

```diff
@@ -0,0 +1,455 @@
+        proc_layer: interfaces.layers.DataLayerInterface,
+        architecture: str,
+        vads: List[Tuple[int, int, str]],
+        address: int,
+    ) -> Optional[Tuple[int, str]]:
+        """
+        Args:
+            syscall_finder:
+            proc_layer: the memory layer of the process being scanned
+            architecture: the name of the architecture for the process being disassembled
+            vads: the ranges of this process under 10MB
+            address: the starting address to check for malicious syscall code blocks
+
+        Returns:
+            Optional[Tuple[int, str]]: For valid code blocks, the starting address of the block and the disassembly string
+        """
+        # the number bytes behind the yara rule hit to scan
+        behind = 32
+
+        address = address - behind
+
+        try:
+            data = proc_layer.read(address, behind * 2)
+        except exceptions.InvalidAddressException:
+            return None
+
+        disasm_func = cls.get_disasm_function(architecture)
+
+        # since Intel does not have fixed-size instructions, we have to scan
+        # each byte offset and re-disassemble the remaining block
+        for offset in range(behind):
+            # if this looks like a system call back (r10, rax, ret/jmp)
+            syscall_info = cls._is_syscall_block(
+                disasm_func, syscall_finder, data[offset:], address + offset
+            )
+            if syscall_info:
+                disasm_bytes, end_inst = syscall_info
+
+                # if we can recover (and require) a target address for this malware technique
+                if syscall_finder.get_syscall_target_address:
+                    target_address = syscall_finder.get_syscall_target_address(
+                        proc_layer, end_inst
+                    )
+
+                    # could not determine the address -> invalid basic block
+                    if not target_address:
+                        continue
+
+                    # we only care about calls to system call DLLs
+                    path = cls.get_range_path(vads, target_address)
+                    if not isinstance(path, str) or not path.lower().endswith(
+                        cls.valid_syscall_handlers
+                    ):
+                        continue
+
+                # return the address and disassembly string if all checks pass
+                return address + offset, disasm_bytes
+
+        return None
+
+    @staticmethod
+    def get_vad_maps(
+        task: interfaces.objects.ObjectInterface,
+    ) -> List[Tuple[int, int, str]]:
+        """Creates a map of start/end addresses within a virtual address
+        descriptor tree.
+
+        Args:
+            task: The EPROCESS object of which to traverse the vad tree
+
+        Returns:
+            An iterable of tuples containing start and end addresses for each descriptor
+        """
+        vads: List[Tuple[int, int, str]] = []
```

Comment:
Feels like this code will already have been done somewhere else?  I'm happy if you want to duplicate it.  Genuinely not sure which would be the better solution (ie, how specific/responsive to future changes this would be?)

Reply:
There are too many variations of it too really keep one central place. Some plugins want start, size - others start, end - some want only vads with mapped files, others filter to size, etc. Given all the variations, and that `get_vad_root` and `traverse` do all the heavy lifting, it seems best to let each plugin gather whatever ranges it needs.